nemo/collections/tts/modules/transformer_2501.py

```diff
@@ -0,0 +1,704 @@
+                    d_model=d_model,
+                    d_ffn=d_ffn,
+                    sa_n_heads=sa_n_heads,
+                    kernel_size=kernel_size,
+                    p_dropout=p_dropout,
+                    has_xattn=has_xattn,
+                    xa_d_memory=xa_d_memory,
+                    xa_n_heads=xa_n_heads,
+                    is_causal=is_causal,
+                    apply_norm_to_cond=apply_norm_to_cond,
+                    max_length_causal_mask=max_length_causal_mask,
+                    conv_non_linearity=conv_non_linearity,
+                )
+            )
+
+        self.use_learnable_pos_emb = use_learnable_pos_emb
+        self.position_embeddings = None
+        if self.use_learnable_pos_emb:
+            self.position_embeddings = torch.nn.Embedding(max_length_causal_mask, d_model)
+        # Apply random uniform init for all layers, except for output layers: The second of the two layers in the MLP
+        # and the last linear projection in dot product attention. The output layers are scaled depending on the
+        # number of layers
+        self.apply(self._init_weights_gpt2)
+        for name, param in self.named_parameters():
+            if 'o_net' in name and name.endswith('weight'):
+                torch.nn.init.normal_(param, mean=0.0, std=0.02 / math.sqrt(2 * n_layers))
+
+    def reset_cache(self, use_cache=False):
+        for layer in self.layers:
+            layer.reset_cache(use_cache)
+
+    @staticmethod
+    def _init_weights_gpt2(module):
+        if isinstance(module, (torch.nn.Linear, torch.nn.Embedding)):
+            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+
+        if isinstance(module, torch.nn.Linear) and module.bias is not None:
+            torch.nn.init.zeros_(module.bias)
+
+    @staticmethod
+    def _get_layer_inputs(
+        idx: int,
+        cond: Optional[Union[torch.Tensor, List[torch.Tensor]]],
+        cond_mask: Optional[Union[torch.Tensor, List[torch.Tensor]]],
+        attn_prior: Optional[Union[torch.Tensor, List[torch.Tensor]]],
+        multi_encoder_mapping: Optional[List[Optional[int]]],
+    ) -> Tuple[Optional[torch.Tensor], Optional[torch.Tensor], Optional[torch.Tensor]]:
+        if multi_encoder_mapping is not None:
+            if multi_encoder_mapping[idx] is None:
+                return None, None, None
+            else:
+                return (
+                    cond[multi_encoder_mapping[idx]],
+                    cond_mask[multi_encoder_mapping[idx]] if cond_mask is not None else None,
+                    attn_prior[multi_encoder_mapping[idx]] if attn_prior is not None else None,
+                )
+        else:
+            return cond, cond_mask, attn_prior
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        x_mask: torch.Tensor,
+        cond: Optional[Union[torch.Tensor, List[torch.Tensor]]] = None,
+        cond_mask: Optional[Union[torch.Tensor, List[torch.Tensor]]] = None,
```

Comment:
nit: we now support the types of `cond` either `torch.Tensor` or a list of `torch.Tensor`. Is it better to unify the two as only a list of `torch.Tensor`? in this case, a list of a single tensor corresponds to a single encoder, and a list of multipe tensors corresponds to multiple encoders?

Reply:
@paarthneekhara what is your thoughts on that?